### PR TITLE
string join() renovation

### PR DESCRIPTION
**Pull request by psahabu**
@benharsh:
+ cleaned up the original function, which took an array of strings
+ made it generic over tuples and arrays
+ added varargs, tuple, and array wrapper functions around the generic version

### CODE REVIEW

File: modules/internal/String.chpl [hunk 1]
```diff
@@ -731,52 +731,78 @@ module String {
...
+
+          var x = "|".join("a","10","d");
+          writeln(x); // prints: "a|10|d"
+     */
+
+    proc join(S: string ...) : string {
+      return _join(S);
+    }
+
+    proc join(S) : string where isTuple(S) {
```
Comment (benharsh) [inline modules/internal/String.chpl]: I think we could add some chpldoc text here as well.

File: modules/internal/String.chpl [hunk 2]
```diff
@@ -731,52 +731,78 @@ module String {
...
-        ret.len = newSize;
-        const allocSize = chpl_here_good_alloc_size(ret.len+1);
-        ret._size = allocSize;
-        ret.buff = chpl_here_alloc(allocSize,
-                                  CHPL_RT_MD_STR_COPY_DATA): bufferType;
-        var offset = 0;
+      return _join(S);
+    }
+
+    proc _join(const ref S) : string where isTuple(S) || isArray(S) {
```
Comment (benharsh) [inline modules/internal/String.chpl]: Should this be private?

Comment (psahabu) [inline modules/internal/String.chpl]: I tried but I got this error when running my tests:

$CHPL_HOME/modules/internal/String.chpl:773: error: Can't apply private to the fields or methods of a class or record yet

**Comment by benharsh** [comment]
The rest looks good to me 👍

**Pull request merged**